### PR TITLE
Use "rejection sampling" for generating field elements

### LINKED ISSUE #13: Specify method for generating field elements (was "Proposal: use hash to field to generate field elements from a seed")
> TODO This functionality closely resembles what people usually think of as an extract-then-expand KDF, but differs somewhat in its syntax and, also, its required security properties. Can we get the same functionality from something that's more commonplace? HKDF doesn't fit the bill, unfortunately, because keys can only be expanded to a fairly short length. Our application requires a rather long key stream.

and
>    *  Field.rand_vec(len: Unsigned) -> output: Vec[Field] returns a
      vector of random field elements.  The length of output MUST be
      len.
>
>         NOTE In reality this would be achieved by generating a random
         key and expanding it into a sequence of field elements using a
         key derivation scheme.  This should probably be made explicit.

In hash to curve draft, there is a method to derive an arbitrary number of field elements from a seed source.
This is called a Expander, and there are two types: one based on Merkle-Damgard functions and the other based on eXtendable Output functions.
See section 5.3 : https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-hash-to-curve#section-5.3

### PR DESCRIPTION
**Pull request by cjpatton**
Closes #13 by doing "rejection sampling". This an alternative to #29 and #30.

### CODE REVIEW

File: draft-patton-cfrg-vdaf.md [hunk 1]
```diff
@@ -839,20 +839,24 @@ def vec_add(left, right):
...
 A concrete `Prg` implements the following class method:
 
-* `Prg.expand(seed: Bytes, info: Bytes, length: Unsigned) -> Bytes` derives a
-  byte string from the given seed and info string. The length of the output MUST
-  be `length`.
+* `Prg(seed: Bytes, info: Bytes, length: Unsigned)` constructs an instance of
+  `Prg` from the given seed and info string.
+
+* `prg.next(length: Unsigned)` returns the next `length` bytes of output of the
+  PRG.
```
Comment (schoppmp) [inline draft-patton-cfrg-vdaf.md:859]: Please add some info about what we expect from the output, e.g.
`
The bytes returned are uniformly pseudorandom when the PRG was constructed with a uniformly random seed.
`

File: draft-patton-cfrg-vdaf.md [hunk 2]
```diff
@@ -861,15 +865,21 @@ pseudorandom field elements.
 ~~~
 # Derive a new seed.
```
Comment (schoppmp) [inline draft-patton-cfrg-vdaf.md:866]: Can you add some information for why this is needed? Also please add that the seed MUST have high entropy.

File: draft-patton-cfrg-vdaf.md [hunk 3]
```diff
@@ -861,15 +865,21 @@ pseudorandom field elements.
...
 def derive(Prg, seed: bytes, info: bytes) -> bytes:
-    return Prg.expand(seed, info, Prg.SEED_SIZE)
+    prg = Prg(seed, info)
+    return prg.next(Prg.SEED_SIZE)
 
 # Expand a seed into a vector of Field elements.
```
Comment (schoppmp) [inline draft-patton-cfrg-vdaf.md:871]: Same as above, please make sure that this is always used in a safe way by requiring that the seed MUST have high entropy and be only used once.

File: draft-patton-cfrg-vdaf.md (hunk 1 above)
Comment (cjpatton) [reply draft-patton-cfrg-vdaf.md:859]: Done. I also stipulated that the random seed might be an output of a previous call to the PRG.

File: draft-patton-cfrg-vdaf.md (hunk 2 above)
Comment (cjpatton) [reply draft-patton-cfrg-vdaf.md:866]: The intended usage is explained above:
> Each `Prg` has two derived class methods. The first is used to derive a fresh seed from an existing one.

If you want to see this get used, search for `Prg.derive` in the prio3 section.

High entropy ins't going to be sufficient for all PRGs, in particular for those constructed from a pseudorandom permutation (e.g., AES). I stipulated the same requirement for securely generating seeds as above.

File: draft-patton-cfrg-vdaf.md (hunk 3 above)
Comment (cjpatton) [reply draft-patton-cfrg-vdaf.md:871]: It depends on the PRG construction, but it should be secure to reuse the same seed with different "info" strings. That is, the intent of the "info" string is to provide domain separation.

Here I stipulated the same requirements for generating the seed as above.

File: draft-patton-cfrg-vdaf.md (hunk 2 above)
Comment (schoppmp) [reply draft-patton-cfrg-vdaf.md:871]: The current documentation sounds good, thanks!

File: draft-patton-cfrg-vdaf.md [hunk 4]
```diff
@@ -839,37 +839,52 @@ def vec_add(left, right):
...
-  byte string from the given seed and info string. The length of the output MUST
-  be `length`.
+* `Prg(seed: Bytes, info: Bytes, length: Unsigned)` constructs an instance of
+  `Prg` from the given seed and info string. The seed MUST be of length
+  `SEED_SIZE` and MUST be generated securely (i.e., it is either the output of
+  `gen_rand` or a previous invocation of the PRG).
+
+* `prg.next(length: Unsigned)` returns the next `length` bytes of output of the
+  PRG. If the seed was securely generated, the output can be treated as
+  pseudorandom.
```
Comment (schoppmp) [inline draft-patton-cfrg-vdaf.md:862]: Maybe add a note about `info` and domain separation?

File: draft-patton-cfrg-vdaf.md (hunk 3 above)
Comment (schoppmp) [reply draft-patton-cfrg-vdaf.md:876]: Okay, let's maybe add a short sentence on this to the documentation (see comment above).

**Review by schoppmp — APPROVED** [review]

File: draft-patton-cfrg-vdaf.md (hunk 4 above)
Comment (cjpatton) [reply draft-patton-cfrg-vdaf.md:862]: Done

**Pull request merged**